src/Cache/Cache.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TwigCsFixer\Cache;
+
+use InvalidArgumentException;
+use TwigCsFixer\RuleSet\RuleSet;
+use UnexpectedValueException;
+
+/**
+ * @internal
+ */
+final class Cache implements CacheInterface
+{
+    /**
+     * @var SignatureInterface
+     */
+    private $signature;
+
+    /**
+     * @var array<string, int>
+     */
+    private $hashes = [];
+
+    public function __construct(SignatureInterface $signature)
+    {
+        $this->signature = $signature;
+    }
+
+    public function getSignature(): SignatureInterface
+    {
+        return $this->signature;
+    }
+
+    public function has(string $file): bool
+    {
+        return \array_key_exists($file, $this->hashes);
+    }
+
+    public function get(string $file): ?int
+    {
+        if (!$this->has($file)) {
+            return null;
+        }
+
+        return $this->hashes[$file];
+    }
+
+    public function set(string $file, int $hash): void
+    {
+        $this->hashes[$file] = $hash;
+    }
+
+    public function clear(string $file): void
+    {
+        unset($this->hashes[$file]);
+    }
+
+    public function toJson(): string
+    {
+        $json = json_encode([
+            'php'     => $this->getSignature()->getPhpVersion(),
+            'version' => $this->getSignature()->getFixerVersion(),
+            'sniffs'  => array_keys($this->getSignature()->getRuleset()->getSniffs()),
+            'hashes'  => $this->hashes,
+        ]);
+
+        if (\JSON_ERROR_NONE !== json_last_error()) {
```

Comment:
I prefer to use the flag `\JSON_THROW_ON_ERROR` when encoding. (We can still catch JSONException to improve the error message if needed)